### PR TITLE
Upgrade WirePattern to speak IP objects

### PR DESCRIPTION
**Pull request by trustmaster**
#429

### CODE REVIEW

**Comment by trustmaster** [comment]
@bergie feeling brave to review this?

File: spec/Helpers.coffee [hunk 1]
```diff
@@ -66,25 +66,32 @@ describe 'Component traits', ->
...
-      c.inPorts.z.attach z
-      c.outPorts.point.attach p
+      c = null
+      x = null
+      y = null
+      z = null
+      p = null
+      beforeEach (done) ->
+        c = new noflo.Component
+        c.inPorts.add 'x',
```
Comment (bergie) [inline spec/Helpers.coffee]: Why are you instantiating in every test now? We should make sure components are safe for multiple runs

Comment (trustmaster) [inline spec/Helpers.coffee]: Changed this while debugging to put state out of the way. Can make it `beforeEach` => `before` again.

File: src/lib/Helpers.coffee [hunk 2]
```diff
@@ -424,59 +463,67 @@ exports.WirePattern = (component, config, proc) ->
...
-                outs[name] = component.outPorts[name]
+                outs[name] = wrp
 
             outs = outs[outPorts[0]] if outPorts.length is 1 # for simplicity
             groups = [] unless groups
+            # Filter empty connect/disconnect groups
+            groups = (g for g in groups when g isnt null)
             whenDoneGroups = groups.slice 0
             whenDone = (err) ->
+              # TODO add scope to error output
```
Comment (bergie) [inline src/lib/Helpers.coffee]: Scope handling is pretty crucial. Can we get this into the PR?

Comment (trustmaster) [inline src/lib/Helpers.coffee]: Good catch. Simply forgot this.

File: src/lib/Helpers.coffee [hunk 3]
```diff
@@ -100,12 +116,16 @@ exports.WirePattern = (component, config, proc) ->
...
+  if config.receiveStreams
+    throw new Error 'WirePattern receiveStreams is deprecated'
+  # if typeof config.receiveStreams is 'string'
+  #   config.receiveStreams = [ config.receiveStreams ]
   # Send streams feature
   config.sendStreams = false unless 'sendStreams' of config
-  if typeof config.sendStreams is 'string'
-    config.sendStreams = [ config.sendStreams ]
+  if config.sendStreams
+    throw new Error 'WirePattern sendStreams is deprecated'
```
Comment (bergie) [inline src/lib/Helpers.coffee]: This needs a CHANGES entry

File: src/lib/Helpers.coffee [hunk 4]
```diff
@@ -100,12 +116,16 @@ exports.WirePattern = (component, config, proc) ->
   config.forwardGroups = false unless 'forwardGroups' of config
   # Receive streams feature
   config.receiveStreams = false unless 'receiveStreams' of config
-  if typeof config.receiveStreams is 'string'
-    config.receiveStreams = [ config.receiveStreams ]
+  if config.receiveStreams
+    throw new Error 'WirePattern receiveStreams is deprecated'
```
Comment (bergie) [inline src/lib/Helpers.coffee]: This needs a CHANGES entry

File: src/lib/Helpers.coffee [hunk 5]
```diff
@@ -100,12 +116,16 @@ exports.WirePattern = (component, config, proc) ->
   config.forwardGroups = false unless 'forwardGroups' of config
```
Comment (bergie) [inline src/lib/Helpers.coffee]: I wonder if we should swap the default now to `true`, just like `forwardBrackets`

File: spec/Helpers.coffee (hunk 1 above)
Comment (bergie) [inline spec/Helpers.coffee]: Attaching output socket in `beforeEach`, and detaching in `afterEach` is fine, however

**Comment by bergie** [comment]
I'm wondering also if we should deprecate `async: false`

File: src/lib/Helpers.coffee (hunk 3 above)
Comment (trustmaster) [inline src/lib/Helpers.coffee]: ✅

File: src/lib/Helpers.coffee (hunk 4 above)
Comment (trustmaster) [inline src/lib/Helpers.coffee]: ✅

File: src/lib/Helpers.coffee (hunk 5 above)
Comment (trustmaster) [inline src/lib/Helpers.coffee]: That will break some things. I'd like to keep backwards compatibility layer as backwards compatible as possible.

**Comment by trustmaster** [comment]
@bergie added scopes to errors and updated per other notes.

Regarding deprecating `async: false` and #370, are we going to backport features to WirePattern until 2.0?

**Pull request merged**